src/main/java/com/hackathon03/codeguru/utils/ConcurrencyDeadlock.java

```diff
@@ -0,0 +1,27 @@
+package com.hackathon03.codeguru.utils;
+
+public class ConcurrencyDeadlock {
+
+    Object syncObject1 = new Object();
+    Object syncObject2 = new Object();
+
+    // Noncompliant: both methods request the same two locks in the opposite order, which can cause deadlock.
+
+    public void noncompliantsync1() {
+        synchronized (syncObject1) {
+            synchronized (syncObject2) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 The code acquires two synchronization objects in **two different orders**, which can cause **circular wait** (deadlock).

 The **first order** is: first acquire `syncObject1` and then `syncObject2`:

- enter `noncompliantsync1()` at** line 10**

- **acquire synchronization** on `syncObject1` at** line 11**

- **acquire synchronization** on `syncObject2` at** line 12**



 The **second order** is: first acquire `syncObject2` and then `syncObject1`:

- enter `noncompliantsync2()` at** line 19**

- **acquire synchronization** on `syncObject2` at** line 20**

- **acquire synchronization** on `syncObject1` at** line 21**



**Fix**  
 The fix is to acquire synchronization in only one order. Typically you can do this by adding extra synchronization to the synchronization order you want to prohibit.